### PR TITLE
kubeadm: do not set deprecated '--cgroup-driver' flag in kubeadm-flags.env, it will be set in config.yaml

### LINKED ISSUE #949: kubeadm-flags.env contains deprecated kubelet flags
EDIT by neolit123:
help-wanted

k/k issue of migrating flags:
https://github.com/kubernetes/kubernetes/issues/86843

the flags in this file have to be moved to KubeletConfiguration instead.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/flags.go

the kubelet supports some flags as config fields and we should use them instead because the flags are deprecated:
https://godoc.org/k8s.io/kubelet/config/v1beta1#KubeletConfiguration

buildKubeletArgMap needs to include only flags that are unique and are not present in the config.

--------

# BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0-rc.1", GitCommit:"8745ea56e3f1f3ad20050c1762eb6ba6f7786675", GitTreeState:"clean", BuildDate:"2018-06-20T23:58:14Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0-rc.1", GitCommit:"8745ea56e3f1f3ad20050c1762eb6ba6f7786675", GitTreeState:"clean", BuildDate:"2018-06-21T00:01:04Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
- **Cloud provider or hardware configuration**:
Brightbox
- **OS** (e.g. from /etc/os-release):
Ubuntu 18.04 LTS
- **Kernel** (e.g. `uname -a`):
Linux srv-fo4d2 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- **Others**:

## What happened?
kubeadm init bootstrap sets both the `--resolv-conf` flag in `kubeadm-flags.env` and the `resolvConf` entry in `config.yaml`

And to different entries: the flag to the external resolver file, and the yaml entry to the systemd stub resolver.

This generates a warning in the journal.

## What you expected to happen?

Just use the resolvConf in the yaml file and decide which resolver to use :-)

## How to reproduce it (as minimally and precisely as possible)?

kubeadm init, and stop it during the certificate generation (would use `alpha phase preflight master` but it is still broken with flags and config files).

## Anything else we need to know?
Using containerd

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
kubeadm: do not set deprecated '--cgroup-driver' flag in kubeadm-flags.env, this value will be set in config.yaml

```
--cgroup-driver string  Driver that the kubelet uses to manipulate cgroups on the host.  Possible values: 'cgroupfs', 'systemd' (default "cgroupfs") (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)

```

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#949

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
ACTION REQUIRED: kubeadm does not set the deprecated '--cgroup-driver' flag in /var/lib/kubelet/kubeadm-flags.env, it will be set in the kubelet config.yaml. If you have this flag in /var/lib/kubelet/kubeadm-flags.env or /etc/default/kubelet (/etc/sysconfig/kubelet for RPMs) please remove it and set the value using KubeletConfiguration
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by neolit123** [comment]
thank you @SataQiu this looks clean to me.
/lgtm
/assign @rosti 
for approval.

but i think we need an action required this time and here is why.
if the users are upgrading from e.g. 1.18 to 1.19 and the 1.19 kubelet no longer has flags such as `--cgroup-driver` or `--resolv-conf` it will fail executing. kubeadm writes these flags on disk in:
`/var/lib/kubelet/kubeadm-flags.env`
and the user can override with:
`/etc/default/kubelet`

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/#the-kubelet-drop-in-file-for-systemd

the problem is these files persist on disk and kubeadm upgrade does not touch them.
so in my opinion the action required by the user is to remove these flags manually from the above files.

suggesting change:

> ACTION REQUIRED: kubeadm: do not set the deprecated '--cgroup-driver' flag in /var/lib/kubelet/kubeadm-flags.env, it will be set in the kubelet config.yaml. If you have this flag in /var/lib/kubelet/kubeadm-flags.env or /etc/default/kubelet please remove it and set the value using KubeletConfiguration

please also update the release note in this PR:
https://github.com/kubernetes/kubernetes/pull/90394

> ACTION REQUIERD: kubeadm: fix the bug that kubeadm does not really respect resolvConf value set by user if systemd-resolved service is active. kubeadm no longer sets the flag in /var/lib/kubelet/kubeadm-flags.env. If you have this flag in /var/lib/kubelet/kubeadm-flags.env or /etc/default/kubelet please remove it and set the value using KubeletConfiguration

**Comment by neolit123** [comment]
and we need to update this in the docs:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#configure-cgroup-driver-used-by-kubelet-on-control-plane-node

File: cmd/kubeadm/app/util/config/initconfiguration.go [hunk 1]
```diff
@@ -141,9 +141,9 @@ func SetAPIEndpointDynamicDefaults(cfg *kubeadmapi.APIEndpoint) error {
 }
 
 // SetClusterDynamicDefaults checks and sets values for the ClusterConfiguration object
-func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, LocalAPIEndpoint *kubeadmapi.APIEndpoint) error {
+func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint *kubeadmapi.APIEndpoint, nodeRegOpts *kubeadmapi.NodeRegistrationOptions) error {
```
Comment (rosti) [inline cmd/kubeadm/app/util/config/initconfiguration.go:144]: Note to myself: This looks fairly awkward here and is a prominent reminder that component configs should not be housed in the ClusterConfiguration as they are no longer cluster wide.

**Review by rosti — APPROVED** [review]
Thanks @SataQiu !
/approve

**Pull request merged**